### DISCUSSION TITLE
issue: incompatible types for `control`

### DISCUSSION THREAD

**Discussion by scallaway** [react-hook-form/react-hook-form#9102] [Q&A]
### Version Number

7.36.1

### Codesandbox/Expo snack

https://codesandbox.io/s/priceless-gagarin-5hww6t?file=/src/App.tsx

### Steps to reproduce

Try to implement a generic component that should accept `control` in order to know the form that it's associated with.

Unfortunately, the Type Checking doesn't pass since the types for `'_options.resolver'` are incompatible.

### Expected behaviour

You should be able to extend `FieldValues` with the generic so that the component that you build can be sufficiently generic for your application.

Given `FieldValues` is aliased to `Record<string, any>` this should be perfectly doable. However, with the current typing, this doesn't seem to work.

There are multiple other issues that have been created for this, but they've all been closed whilst this issue still persists with the latest version. Therefore, I'm hoping this will produce some good conversation again.

---

TS 4.8
React 17

### Relevant log output

```shell
'import("/sandbox/node_modules/react-hook-form/dist/types/form").Control<FieldValues, any>' is not assignable to type 'import("/sandbox/node_modules/react-hook-form/dist/types/form").Control<import("/sandbox/node_modules/react-hook-form/dist/types/fields").FieldValues, any>'.
  The types of '_options.resolver' are incompatible between these types.
    Type 'import("/sandbox/node_modules/react-hook-form/dist/types/resolvers").Resolver<FieldValues, any> | undefined' is not assignable to type 'import("/sandbox/node_modules/react-hook-form/dist/types/resolvers").Resolver<import("/sandbox/node_modules/react-hook-form/dist/types/fields").FieldValues, any> | undefined'.
      Type 'import("/sandbox/node_modules/react-hook-form/dist/types/resolvers").Resolver<FieldValues, any>' is not assignable to type 'import("/sandbox/node_modules/react-hook-form/dist/types/resolvers").Resolver<import("/sandbox/node_modules/react-hook-form/dist/types/fields").FieldValues, any>'.ts(2322)
InputField.tsx(6, 12): The expected type comes from property 'control' which is declared here on type 'IntrinsicAttributes & InputFieldProps<FieldValues>'
`
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
issue to due to memo type inference: https://codesandbox.io/s/distracted-sunset-cmlbpc?file=/src/App.tsx

**Reply by scallaway** [reply]
Oh wow, that's annoyingly simple.

Although it seems rather unsatisfactory that I can't use memoisation with `control` like this.

Do you know if this is a problem with the memoisation typing or something in this library?

**Reply by bluebill1049** [reply]
honestly, i don't see much of the value with memo, since the component re-render is already getting isolated by hook form.

**Reply by scallaway** [reply]
Yeah fair - I'm also seeing this with `forwardRef` as well, when trying to pass `ref`s down to functional components.

**Reply by scallaway** [reply]
Yeah I don't think simply removing the memoisation is the answer to this question.

Given I'm seeing the same problems with `React.forwardRef`, without using any memoisation at all. This is definitely a problem when working with MUI as certain components require their children to accept `ref`s.

The _workaround_ is obviously to move those components (where possible) inside the `<Controller />` but that's not always feasible.

**Reply by bluebill1049** [reply]
This type issue is not related to this library, that's fine if my answer doesn't answer your problem. Mark as unanswered.
